src/blind.rs

```diff
@@ -239,37 +239,161 @@ impl TxOutSecrets {
     /// Returns a tuple (assetid, blind_factor, generator) if the blinds are
     /// consistent with asset commitment
     /// Otherwise, returns an error
-    pub(crate) fn surjection_inputs<C: Signing>(
-        secret: Option<&Self>,
-        secp: &Secp256k1<C>,
-        asset: Asset,
-    ) -> Result<(Generator, Tag, Tweak), TxOutError>
+    pub fn surjection_inputs<C: Signing>(&self, secp: &Secp256k1<C>)
+        -> (Generator, Tag, Tweak)
     {
-        let gen = asset.into_asset_gen(secp)
-            .ok_or(TxOutError::UnExpectedNullAsset)?;
-        match secret {
-            None => {
-                let tag = Tag::from([0u8; 32]);
-                Ok((gen, tag, ZERO_TWEAK))
-            }
-            Some(secret) => {
-                let tag = secret.asset.into_tag();
-                let bf = secret.asset_bf.into_inner();
-                let gen1 = Generator::new_blinded(secp, tag, bf);
-                if gen1 != gen {
-                    return Err(TxOutError::IncorrectBlindingFactors);
-                }
-                Ok((gen, tag, bf))
-            }
-        }
+        let tag = self.asset.into_tag();
+        let bf = self.asset_bf.into_inner();
+        let gen = Generator::new_blinded(secp, tag, bf);
+        (gen, tag, bf)
     }
 
     /// Gets the required fields for last value blinding factor calculation from [`TxOutSecrets`]
-    pub(crate) fn value_blind_inputs(&self) -> (u64, AssetBlindingFactor, ValueBlindingFactor) {
+    pub fn value_blind_inputs(&self) -> (u64, AssetBlindingFactor, ValueBlindingFactor) {
         return (self.value, self.asset_bf, self.value_bf);
     }
 }
 
+/// Data structure used to provide inputs to [`SurjectionProof`] methods.
+/// Inputs for which we don't know the secrets can be [`SurjectionInput::Unknown`],
+/// while inputs from user's wallet should be [`SurjectionInput::Known`]
+///
+/// Explicit assets can be provided as [`SurjectionInput::Unknown`]. There is no
+/// need to construct a `Known` variant with secrets
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum SurjectionInput{
+    /// Unknown inputs for whom we don't know the secrets(asset tags/blinding factors)
+    Unknown(Asset),
+    /// Known inputs for whom we know corresponding [`TxOutSecrets`]
+    Known(TxOutSecrets),
```

Comment:
In 8b41c1327d8af43f2196ce97d230c5a0719d74c3:

I think this should be
```
Known {
    asset: AssetId,
    asset_bf: AssetBlindingFactor,
}
```
These fields are both `Copy` and small-ish and easy to copy out of `TxOutSecrets`. This would make this type tighter -- right now you use `TxOutSecrets` and sometimes just put dummy data into them which is potentially confusing.